Refactor use of Config into its own module

thisisshi opened the pull request:
Refactor to consolidate references to `Config` and `Bag` to its own top level module, `c7n.config` in order to better align validate/test/run.

kapilt commented on c7n/commands.py:
```diff
@@ -190,7 +191,7 @@ def validate(options):
         used_policy_names = used_policy_names.union(conf_policy_names)
         if not errors:
-            null_config = Bag(dryrun=True, log_group=None, cache=None, assume_role="na")
+            null_config = set_null_config()
```
we use bag for config in lots of places, i'd rather not introduce a new differential path that needs updating in sync with options, the easier fix is just call Bag.empty(**kw)

thisisshi replied:
Gotcha. Was also looking at tests/common.py, is there a reason behind having two implementations of Bag? Shouldn't the Config class be moved into utils?

kapilt replied:
other way around, bag should probably be used by tests as well. I think it pre-dates the bag implementation.

kapilt commented:
so its worth pulling up and reviewing goals before impl

 - unify usage of config to a single implementation in both app code and tests
 - limit aws specitivity as we work to enable multi-cloud
 - allow for future goals already extant on some of the issues, namely specifying config in policy files.

now wrt to the implementation

 - let's move config into a new top level module config.py
 - note there's another implementation in handler.py that should also use config.py
 - audit existing usage of bag when its actually being used as config
 - separate initializer in tests for config.py that does a binding for functools.partial with test values/variables.

how does that sound?

thisisshi commented:
Sounds good, I'll move Config/Bag out of utils and into its own module and base its implementation off of the existing one in handler.py and handle init for tests in there as well and take a look at the existing imports of Bag/Config inside of tools

thisisshi commented:
Updated with removal of `Config` and `Bag` from handler and moved into its own module c7n.config. I'm not too familiar with working with partials, let me know if the implementation is off. Also, Drone seems to be failing to get the latest commit with c7n.config in it but Travis is running as expected

kapilt commented on c7n/config.py:
```diff
@@ -0,0 +1,69 @@
+
+from c7n.utils import get_account_id_from_sts
+
+logging.root.setLevel(logging.DEBUG)
```
logging level shouldn't be set at module level

kapilt commented on c7n/config.py:
```diff
@@ -0,0 +1,69 @@
+            'cache_period': 0,
+            'dryrun': False})
+        d.update(kw)
+        if not os.path.exists(d['output_dir']):
```
this part shouldn't be here, this should be moved to original caller in handler, ie config construction shouldn't be doing an implicit disk mod.

thisisshi replied:
yeah, makes sense.

kapilt commented on c7n/config.py:
```diff
@@ -0,0 +1,69 @@
+    @classmethod
+    def empty(cls, **kw):
+        account_id = None
+        if 'AWS_LAMBDA_FUNCTION_NAME' in os.environ:
```
this part should be moved to the original caller in handler.py and then passed as parameter to empty()

thisisshi commented:
@kapilt updated to instantiate `Config.empty()` in cli and passed into respective commands

kapilt commented on c7n/config.py:
```diff
@@ -0,0 +1,54 @@
+        d.update(kw)
+        if not os.path.exists(d['output_dir']):
+            try:
+                os.mkdir(d['output_dir'])
```
per previous comments, directory creation for outputs should be moved to the caller, in this context handler.py
